### PR TITLE
Draft: [dalgona] Python interface for dynamic analysis of DNN

### PR DESCRIPTION
**Pull request by jinevening**
On going draft to introduce dalgona, a dynamic analysis tool for DNN

ONE-DCO-1.0-Signed-off-by: Hyukjin Jeong <hj1.jeong@samsung.com>

---
Related to https://github.sec.samsung.net/STAR/nnfw/issues/11537

### CODE REVIEW

File: compiler/dalgona/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,38 @@
+cmake_minimum_required(VERSION 3.0)
+project(dalgona)
+
+find_package(PythonLibs REQUIRED)
+if(NOT PythonLibs_FOUND)
+  message(STATUS "Build dalgona: FAILED (missing PythonLibs. try install python3-dev)")
+  set(Pybind11_FOUND FALSE PARENT_SCOPE)
```
Comment (mhs4670go) [inline compiler/dalgona/CMakeLists.txt:7]: This `set` command is needed? And, maybe `return ()` command missing.

Comment (jinevening) [reply compiler/dalgona/CMakeLists.txt]: line 6~8 seems unnecessary. I've removed them. Thanks.

File: infra/cmake/modules/ExternalBuildTools.cmake [hunk 2]
```diff
@@ -60,6 +60,8 @@ function(ExternalBuild_CMake)
                   RESULT_VARIABLE BUILD_EXITCODE)
```
Comment (mhs4670go) [inline infra/cmake/modules/ExternalBuildTools.cmake:60]: maybe this file is added mistakenly?

File: compiler/dalgona/src/HDF5Importer.h [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  {
+    // Do nothing
+  }
+
+public:
+  /**
+   * @brief importGroup has to be called before readTensor is called
+   *        Otherwise, readTensor will throw an exception
+   */
+  void importGroup() { _value_grp = _file.openGroup("value"); }
```
Comment (mhs4670go) [inline compiler/dalgona/src/HDF5Importer.h:52]: (optional) I'm not sure(need to test) but data or group can be accessed by its absolute path as well. Maybe absolute path makes this line be removed and prevent from mistakes of developers. What I think is like
```cxx
void HDF5Importer::readTensor(int32_t record_idx, int32_t input_idx, void *buffer)
{
  // auto record = _value_grp.openGroup(std::to_string(record_idx));
  auto tensor = record.openDataSet("value" + std::to_string(record_idx) + std::to_string(input_idx));

  readTensorData(tensor, static_cast<uint8_t *>(buffer));
}
```

File: infra/cmake/modules/ExternalBuildTools.cmake (hunk 2 above)
Comment (jinevening) [reply infra/cmake/modules/ExternalBuildTools.cmake:60]: It seems that the error occurs because of building tests in pybind11.

```
20:31:52   CMake Error at tests/CMakeLists.txt:197 (message):
20:31:52 
20:31:52     Running the tests requires pytest.  Please install it manually (try:
20:31:52     /usr/bin/python3.5 -m pip install pytest)
20:31:52 
20:31:52   
20:31:52 
20:31:52   
20:31:52 
20:31:52   -- Configuring incomplete, errors occurred!
```

I'm trying to turn off the option to build pybind11 tests.

File: compiler/dalgona/src/HDF5Importer.h (hunk 3 above)
Comment (jinevening) [reply compiler/dalgona/src/HDF5Importer.h:52]: It looks much better. I've changed readTensor. Thanks.

File: infra/cmake/modules/ExternalBuildTools.cmake (hunk 2 above)
Comment (jinevening) [reply infra/cmake/modules/ExternalBuildTools.cmake]: I could prevent building pybind11 tests by adding EXTRA_OPTS as below.
```
  nnas_include(ExternalBuildTools)
  ExternalBuild_CMake(CMAKE_DIR   ${Pybind11Source_DIR}
                      BUILD_DIR   ${CMAKE_BINARY_DIR}/externals/PYBIND11/build
                      INSTALL_DIR ${EXT_OVERLAY_DIR}
                      IDENTIFIER  "2.3.0"
                      PKG_NAME    "PYBIND11"
                      EXTRA_OPTS "-DPYBIND11_TEST:BOOL=OFF")
```

File: compiler/dalgona/src/HDF5Importer.h (hunk 3 above)
Comment (mhs4670go) [reply compiler/dalgona/src/HDF5Importer.h:51]: Maybe this method can be removed now?

File: compiler/dalgona/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,34 @@
+cmake_minimum_required(VERSION 3.0)
+project(dalgona)
```
Comment (seanshpark) [inline compiler/dalgona/CMakeLists.txt:2]: Is this necessary here? (I'm also not good at cmake but this `procject` appeared in top cmake file...)

File: compiler/dalgona/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+find_package(PythonLibs REQUIRED)
```
Comment (seanshpark) [inline compiler/dalgona/CMakeLists.txt:4]: maybe this be changed to `nnas_find_package` like others in below and then if not found, show message - return.

File: compiler/dalgona/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+target_include_directories(dalgona PRIVATE ${Pybind11_INCLUDE_DIRS})
+target_include_directories(dalgona PRIVATE ${HDF5_INCLUDE_DIRS})
+
+target_link_libraries(dalgona INTERFACE pybind11::embed)
+target_link_libraries(dalgona PRIVATE ${HDF5_CXX_LIBRARIES})
+target_link_libraries(dalgona PRIVATE ${PYTHON_LIBRARIES})
+target_link_libraries(dalgona PRIVATE arser)
+target_link_libraries(dalgona PRIVATE safemain)
+target_link_libraries(dalgona PRIVATE luci_import)
+target_link_libraries(dalgona PRIVATE luci_interpreter)
```
Comment (seanshpark) [inline compiler/dalgona/CMakeLists.txt:34]: Do you have no plan for testing? No testing will decrease our test coverage.

**Comment by seanshpark** [comment]
As the PR is quite huge, it'll take time to review all the codes.
You can wait till the full review is finished: don't know how long I'll take,
or you can split so that reviewers can concentrate that may speed up each PRs.

File: compiler/dalgona/README.md [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+## Prerequisite
+- Python (apt install python-dev python3-dev)
+- Circle model (target to analyze)
+- Input data of the model (hdf5 format. See imgdata2hdf5 or gen_h5_explicit_inputs.py for more details.)
+- Analysis code (Python code)
+
+## Example
+```
+dalgona \
+> --input_model model.circle
```
Comment (seanshpark) [inline compiler/dalgona/README.md:26]: It would be nice if you remove `>` at the beginning so that users can try with copy-paste with this example.

File: compiler/dalgona/README.md [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+> --input_data data.hdf5
+> --analysis analysis/AnalysisTemplate.py
+```
+
+## Arguments
+```
+  --help            Show help message and exit
+  --input_model     Input model filepath (.circle)
+  --input_data      Input data filepath (.h5)
+  --analysis        Analysis code filepath (.py)
```
Comment (seanshpark) [inline compiler/dalgona/README.md:36]: (optional) `analyzer` ?

File: compiler/dalgona/README.md [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+  --analysis_args   (optional) String argument passed to the analysis code
+```
+
+## How to write analysis code?
+
+`dalgona` provides hooks which are called before/after an operator is executed.
+Users can access tensors relevant to the corresponding operator inside the hooks.
+The information of each operator is passed as the arguments of the hook.
+For example, for a Conv2D operator, dalgona provides following hooks.
```
Comment (seanshpark) [inline compiler/dalgona/README.md:45]: We usually used our sub-project or module names as italic, _dalgona_ . You use `dalgona` and dalgona. It would be better to use one convention.

File: compiler/dalgona/README.md [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+```
+  def Conv2DPre(self, name, input, filter, bias, padding, stride, dilation, fused_act)
+  def Conv2DPost(self, name, input, filter, bias, padding, stride, dilation, output, fused_act)
+```
+
+Conv2DPre/Conv2DPost are called before/after Conv2D is executed, respectively. Users can write codes to analyze the distribution of intermediate tensors using the provided arguments.
```
Comment (seanshpark) [inline compiler/dalgona/README.md:52]: (optional) `Conv2DPre`/`Conv2DPost`

File: compiler/dalgona/include/PostOperatorHook.h [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+    auto py_stride = py::dict("w"_a = stride->w(), "h"_a = stride->h());
+    auto py_dilation = py::dict("w"_a = dilation->w(), "h"_a = dilation->h());
+
+    auto fused_act = node->fusedActivationFunction();
+
+    hook(node->name(),                                      // name
+         inputs[0],                                         // input
+         inputs[1],                                         // filter
+         inputs[2],                                         // bias
+         padding == luci::Padding::SAME ? "SAME" : "VALID", // padding
```
Comment (seanshpark) [inline compiler/dalgona/include/PostOperatorHook.h:110]: (optional) use `padding_str = padding == luci::Padding::SAME ? "SAME" : "VALID"` so that arguments and explanation comments get closer.

File: compiler/dalgona/include/PreOperatorHook.h [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+{
+
+// This macro describes common tasks inside pre-opeartor hook
+// After this macro, two variables are created
+// 1. hook: Python function called before operator execution (type: py::object)
+// 2. inputs: input data (type: std::vector of numpy array)
+#define PRE_OPERATOR_HOOK_PROLOGUE(OP_NAME)             \
+  if (!py::hasattr(_analysis, #OP_NAME "Pre"))          \
+  {                                                     \
+    visit(static_cast<const luci::CircleNode *>(node)); \
```
Comment (seanshpark) [inline compiler/dalgona/include/PreOperatorHook.h:49]: use `must_cast` or handle nullptr

File: compiler/dalgona/include/PreOperatorHook.h [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+    return;                                             \
+  }                                                     \
+  py::object hook = _analysis.attr(#OP_NAME "Pre");     \
+  auto inputs = inputsPyArray(node, _interpreter);
+
+private:
+  py::object _analysis;
+  luci_interpreter::Interpreter *_interpreter;
```
Comment (seanshpark) [inline compiler/dalgona/include/PreOperatorHook.h:57]: initialize variable

File: compiler/dalgona/include/OperatorObserver.h [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+  void endNetworkExecution(loco::Graph *graph);
+
+  // Called before an operator is executed
+  void preOperatorExecute(const luci::CircleNode *node) override;
+
+  // Called after an operator is executed
+  void postOperatorExecute(const luci::CircleNode *node) override;
+
+private:
+  luci_interpreter::Interpreter *_interpreter;
```
Comment (seanshpark) [inline compiler/dalgona/include/OperatorObserver.h:65]: initialize variable

File: compiler/dalgona/src/HDF5Importer.h [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+
+  bool isRawData() { return _value_grp.attrExists("rawData"); }
+
+  int32_t numRecords() { return _value_grp.getNumObjs(); }
+
+  int32_t numInputs(int32_t record_idx);
+
+private:
+  H5::H5File _file;
+  H5::Group _value_grp;
```
Comment (seanshpark) [inline compiler/dalgona/src/HDF5Importer.h:76]: maybe need variable initialize?

**Comment by seanshpark** [comment]
There are no unit tests and no negative tests.
- This will decrease coverage and negative test percentage.

Please add tests.

File: compiler/dalgona/include/PreOperatorHook.h (hunk 13 above)
Comment (jinevening) [reply compiler/dalgona/include/PreOperatorHook.h:57]: These variables are initialized when `PreOperatorHook` is constructed.

File: compiler/dalgona/CMakeLists.txt (hunk 4 above)
Comment (jinevening) [reply compiler/dalgona/CMakeLists.txt:2]: > Is this necessary here? 

I don't think so. I removed the code. Thanks.

File: compiler/dalgona/CMakeLists.txt (hunk 5 above)
Comment (jinevening) [reply compiler/dalgona/CMakeLists.txt:4]: `PythonLibs` is not the part of our `infra`, so I used `find_package`. I'll use `nnas_find_package` when `PythonLibs` is ready.

FYI, `common-artifacts` is also using `find_package` for finding `PythonLibs`.

File: compiler/dalgona/include/OperatorObserver.h (hunk 14 above)
Comment (jinevening) [reply compiler/dalgona/include/OperatorObserver.h:65]: _interpreter is initialized when OperatorObserver is constructed. 

`py::object _analysis` is initialized with default constructor, which creates a handle with a `nullptr` pointer. I'll add a test for checking if _analysis is null. It would be better to initialize `_analysis` when OperatorObserver is constructed, but that may require a large change to the code.. I'll leave it as a TODO.

File: compiler/dalgona/src/HDF5Importer.h (hunk 15 above)
Comment (jinevening) [reply compiler/dalgona/src/HDF5Importer.h]: `_file` is initialized when `HDF5Importer` is constructed. I've removed `_value_grp`. Thanks.

File: compiler/dalgona/src/HDF5Importer.h (hunk 3 above)
Comment (jinevening) [reply compiler/dalgona/src/HDF5Importer.h]: I removed `importGroup` and `_value_grp` member variable. Thanks!

File: compiler/dalgona/include/PreOperatorHook.h (hunk 13 above)
Comment (seanshpark) [reply compiler/dalgona/include/PreOperatorHook.h:58]: OK, please check with our static code analysis, SVACE.

File: compiler/dalgona/include/OperatorObserver.h (hunk 14 above)
Comment (seanshpark) [reply compiler/dalgona/include/OperatorObserver.h:65]: I've commented cause our static analysis tools usually warn variables without initialization.

Comment (jinevening) [reply compiler/dalgona/include/OperatorObserver.h:65]: I see. I'll initialize the variables to avoid warnings. Thanks.

**Comment by jinevening** [comment]
@seanshpark , I've added `tests` directory, which checks if operator information is correctly passed to operator hooks.

I know it is very difficult to review a huge PR like this. I'll split this into several PRs, which will be much easier to review. Thanks for the reviews. @seanshpark @mhs4670go

File: compiler/dalgona/CMakeLists.txt (hunk 6 above)
Comment (jinevening) [reply compiler/dalgona/CMakeLists.txt:37]: I've added a `tests` directory for testing operator hooks.

**Comment by jinevening** [comment]
Commits
- [ ] Init (create dir, readme, cmake, driver)
- [ ] Introduce OperatorObserver
- [ ] Introduce Pre/PostOperatorHook
- [ ] Introduce AnalysisTemplate
- [ ] Introduce Tests

**Comment by jinevening** [comment]
@seanshpark I'd like to land this project, because this tool is useful for below purposes.

- Cross-layer equalization (confidential)
- Visualization (https://github.com/Samsung/ONE/issues/9694)

The test fails because our CI workers do not have required libraries (python3-dev, python3.8-dev). I'll start with installing dev libraries on our CI workers.

**Comment by jinevening** [comment]
@nnfw-bot test nncc-debug

**Comment by jinevening** [comment]
@nnfw-bot test nncc-release

**Comment by jinevening** [comment]
Done.
